File: Binary Tree Level Order Traversal II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > levelOrderBottom(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        
        vector<vector<int>> result;
        list<TreeNode *>nodeList;
        nodeList.push_back(root);
        
        TreeNode *curNode = NULL;
        int cur = 1;
        int next = 0;
        vector<int> oneResult;
        
        while(!nodeList.empty())
        {
            curNode=nodeList.front();
            nodeList.pop_front();
            cur--;
           
            if(curNode)
            {
                oneResult.push_back(curNode->val);
                
                if(curNode->left)
                {
                    nodeList.push_back(curNode->left);
                    next++;
                }
                
                if(curNode->right)
                {
                    nodeList.push_back(curNode->right);
                    next++;
                }
                
                if(cur==0)
                {
                    auto it = result.begin();
                    result.insert(it, oneResult);
                    oneResult.clear();
                    cur = next;
                    next = 0;
                }
            }
            
         
        }
        
        return result;
        
    }
};

File: Binary Tree Level Order Traversal II.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<ArrayList<Integer>> levelOrderBottom(TreeNode root) {
        // Start typing your Java solution below
        // DO NOT write main() function
        ArrayList<ArrayList<Integer>> levelList = new ArrayList<ArrayList<Integer>>();
        dfs(root, 1, levelList);
        Collections.reverse(levelList);
        return levelList;
    }
    
    private void dfs(TreeNode root, int level, ArrayList<ArrayList<Integer>> levelList)
    {
        if(root == null)
        {
            return;
        }
        
        if(level > levelList.size())
        {
            levelList.add(new ArrayList<Integer>());
        }
        
        levelList.get(level-1).add(root.val);
        dfs(root.left, level + 1, levelList);
        dfs(root.right, level + 1, levelList);
    }
}